### ISSUE TITLE
running ember test in ci with coverage enabled starts a process that doesn't exit

### ISSUE THREAD

**Issue by chrisvdp** [sglanzer/ember-cli-blanket#61]
When I try to run tests in ci mode with coverage enabled a process is started that never finishes. I get a file called `lcov.dat` that has coverage data in it. The xunit file is also generated when I ctrl+c the process.

Running on OSX 10.10.3, Node v0.12.1, phantomjs 1.9.16, ember-cli 0.2.3

I run `ember test -c ./testem.ci.json` where testem.ci.json is
```
{
    "framework": "mocha",
    "test_page": "tests/index.html?coverage",
    "reporter": "xunit",
    "launch_in_ci": [
        "PhantomJS"
    ],
    "report_file": "target/webui-test-output.xml"
}
```
my blanket-options.js is 
```
/* globals blanket, module */

var options = {
    modulePrefix: "cov-webapp",
    filter: "//.*cov-webapp/.*/",
    antifilter: "//.*(tests|template).*/",
    loaderExclusions: [],
    enableCoverage: true,
    lcovOptions: {
        outputFile: 'target/lcov.dat'
    },
    cliOptions: {
        reporters: ['lcov']
    }
};
if (typeof exports === 'undefined') {
    blanket.options(options);
} else {
    module.exports = options;
}
```

**Comment by jschilli** [comment]
Does it run cleanly without the “report_file” option?  

We’re hooking into the blanket reporting output - not sure if we’ve tested with that combination.

**Comment by chrisvdp** [comment]
Same problem without report_file

**Comment by jschilli** [comment]
can you post `bower ls` and `npm ls` output?

I run much the same config (on ember-cli 0.2.3 with mocha) and haven't seen this.

Also running `DEBUG="*" ember test -c ./testem.ci.json` will show all of the comms between testem and ember middleware.

**Comment by chrisvdp** [comment]
`bower ls`
```
/webui
├── Case#1.2.1
├── blanket#1.1.5
├─┬ ember#1.11.3 (latest is 1.13.0-beta.1)
│ └── jquery#1.11.3 (2.1.4 available)
├─┬ ember-cli-shims#0.0.3
│ └── ember#1.11.3 (1.13.0-beta.1 available)
├── ember-cli-test-loader#0.1.3
├─┬ ember-data#1.0.0-beta.17
│ └── ember#1.11.3 (1.13.0-beta.1 available)
├─┬ ember-data.model-fragments#0.3.0 (latest is 0.3.3)
│ └── ember-data#1.0.0-beta.17
├─┬ ember-load-initializers#0.1.4
│ └── ember#1.11.3
├─┬ ember-localstorage-adapter#0.5.3
│ ├── ember#1.11.3 (1.13.0-beta.1 available)
│ └── ember-data#1.0.0-beta.17
├─┬ ember-mocha#0.6.2 (latest is 0.7.0)
│ ├── chai#1.10.0 (latest is 2.3.0)
│ ├── ember-mocha-adapter#0.3.1
│ └── mocha#2.1.0 (latest is 2.2.5)
├─┬ ember-resolver#0.1.15
│ └── ember#1.11.3 (1.13.0-beta.1 available)
├── ember-shortcuts#0.0.2
├─┬ ember-simple-auth#0.7.3 (latest is 0.8.0-beta.2)
│ └── ember#1.11.3 (1.13.0-beta.1 available)
├── font-awesome#4.3.0
├── inuit.css#954d11d70b
├── jquery#1.11.3 (latest is 2.1.4)
├── jquery-ajax-retry#0.2.7 extraneous
├── loader.js#3.2.0 (latest is 3.2.1)
├── modernizr#2.8.3
├─┬ pretender#0.1.0 (latest is 0.6.0)
│ ├── FakeXMLHttpRequest#1.0.0 (latest is 1.2.0)
│ └── route-recognizer#0.1.5
├── sinon#e-tag:2081e2-2c
├── spin.js#2.0.2 (latest is 2.1.0)
└── vex#2.3.2
```
`npm ls` warnings. I can attach the full tree, but it is crazy long.
```
npm WARN unmet dependency /webui/node_modules/ember-cli/node_modules/express requires debug@'~2.1.3' but will load
npm WARN unmet dependency /webui/node_modules/ember-cli/node_modules/debug,
npm WARN unmet dependency which is version 2.2.0
npm WARN unmet dependency /webui/node_modules/ember-cli/node_modules/morgan requires debug@'~2.1.3' but will load
npm WARN unmet dependency /webui/node_modules/ember-cli/node_modules/debug,
npm WARN unmet dependency which is version 2.2.0
npm WARN unmet dependency /webui/node_modules/ember-cli/node_modules/rimraf requires glob@'^4.4.2' but will load
npm WARN unmet dependency /webui/node_modules/ember-cli/node_modules/glob,
npm WARN unmet dependency which is version 5.0.3
npm WARN unmet dependency /webui/node_modules/ember-cli/node_modules/broccoli/node_modules/connect requires debug@'~2.1.3' but will load
npm WARN unmet dependency /webui/node_modules/ember-cli/node_modules/debug,
npm WARN unmet dependency which is version 2.2.0
npm WARN unmet dependency /webui/node_modules/ember-cli/node_modules/broccoli/node_modules/connect/node_modules/finalhandler requires debug@'~2.1.3' but will load
npm WARN unmet dependency /webui/node_modules/ember-cli/node_modules/debug,
npm WARN unmet dependency which is version 2.2.0
npm WARN unmet dependency /webui/node_modules/ember-cli/node_modules/express/node_modules/finalhandler requires debug@'~2.1.3' but will load
npm WARN unmet dependency /webui/node_modules/ember-cli/node_modules/debug,
npm WARN unmet dependency which is version 2.2.0
npm WARN unmet dependency /webui/node_modules/ember-cli/node_modules/express/node_modules/send requires debug@'~2.1.3' but will load
npm WARN unmet dependency /webui/node_modules/ember-cli/node_modules/debug,
npm WARN unmet dependency which is version 2.2.0
npm WARN unmet dependency /webui/node_modules/ember-cli-coffeescript/node_modules/ember-cli/node_modules/express requires debug@'~2.1.1' but will load
npm WARN unmet dependency /webui/node_modules/ember-cli-coffeescript/node_modules/ember-cli/node_modules/debug,
npm WARN unmet dependency which is version 2.2.0
npm WARN unmet dependency /webui/node_modules/ember-cli-coffeescript/node_modules/ember-cli/node_modules/morgan requires debug@'~2.1.1' but will load
npm WARN unmet dependency /webui/node_modules/ember-cli-coffeescript/node_modules/ember-cli/node_modules/debug,
npm WARN unmet dependency which is version 2.2.0
npm WARN unmet dependency /webui/node_modules/ember-cli-coffeescript/node_modules/ember-cli/node_modules/broccoli/node_modules/connect requires debug@'~2.1.1' but will load
npm WARN unmet dependency /webui/node_modules/ember-cli-coffeescript/node_modules/ember-cli/node_modules/debug,
npm WARN unmet dependency which is version 2.2.0
npm WARN unmet dependency /webui/node_modules/ember-cli-coffeescript/node_modules/ember-cli/node_modules/broccoli/node_modules/connect/node_modules/finalhandler requires debug@'~2.1.1' but will load
npm WARN unmet dependency /webui/node_modules/ember-cli-coffeescript/node_modules/ember-cli/node_modules/debug,
npm WARN unmet dependency which is version 2.2.0
npm WARN unmet dependency /webui/node_modules/ember-cli-coffeescript/node_modules/ember-cli/node_modules/express/node_modules/finalhandler requires debug@'~2.1.1' but will load
npm WARN unmet dependency /webui/node_modules/ember-cli-coffeescript/node_modules/ember-cli/node_modules/debug,
npm WARN unmet dependency which is version 2.2.0
npm WARN unmet dependency /webui/node_modules/ember-cli-coffeescript/node_modules/ember-cli/node_modules/express/node_modules/send requires debug@'~2.1.1' but will load
npm WARN unmet dependency /webui/node_modules/ember-cli-coffeescript/node_modules/ember-cli/node_modules/debug,
npm WARN unmet dependency which is version 2.2.0
npm ERR! missing: github@^0.2.3, required by bower@1.4.1
npm ERR! invalid: debug@2.2.0 /webui/node_modules/ember-cli/node_modules/debug
npm ERR! invalid: glob@5.0.3 /webui/node_modules/ember-cli/node_modules/glob
npm ERR! invalid: debug@2.2.0 /webui/node_modules/ember-cli-coffeescript/node_modules/ember-cli/node_modules/debug
```
`DEBUG="*" ember test -c ./testem.ci.json` finishes with
```
  express:application compile etag weak +6s
  express:application compile query parser extended +1ms
  express:application compile trust proxy false +0ms
  express:application booting in development mode +0ms
  socket.io:server initializing namespace / +0ms
  socket.io:server creating engine.io instance with opts {"path":"/socket.io"} +1ms
  socket.io:server attaching client serving req handler +1ms
  express:router use / query +3ms
  express:router:layer new / +0ms
  express:router use / expressInit +1ms
  express:router:layer new / +0ms
  express:router use / <anonymous> +0ms
  express:router:layer new / +0ms
  express:router use / serveStatic +0ms
  express:router:layer new / +0ms
  express:router use / jsonParser +26ms
  express:router:layer new / +0ms
  express:router use / <anonymous> +4ms
  express:router:layer new / +0ms
  express:router use / logErrors +0ms
  express:router:layer new / +0ms
  express:router:route new / +0ms
  express:router:layer new / +0ms
  express:router:route get / +0ms
  express:router:layer new / +0ms
  express:router:route new /\/([0-9]+)$/ +0ms
  express:router:layer new /\/([0-9]+)$/ +1ms
  express:router:route get /\/([0-9]+)$/ +0ms
  express:router:layer new / +0ms
  express:router:route new /testem.js +0ms
  express:router:layer new /testem.js +0ms
  express:router:route get /testem.js +0ms
  express:router:layer new / +0ms
  express:router:route new /^\/(?:[0-9]+)(\/.+)$/ +0ms
  express:router:layer new /^\/(?:[0-9]+)(\/.+)$/ +0ms
  express:router:route get /^\/(?:[0-9]+)(\/.+)$/ +0ms
  express:router:layer new / +0ms
  express:router:route new /^\/(?:[0-9]+)(\/.+)$/ +0ms
  express:router:layer new /^\/(?:[0-9]+)(\/.+)$/ +0ms
  express:router:route post /^\/(?:[0-9]+)(\/.+)$/ +0ms
  express:router:layer new / +0ms
  express:router:route new /^(.+)$/ +0ms
  express:router:layer new /^(.+)$/ +0ms
  express:router:route get /^(.+)$/ +1ms
  express:router:layer new / +0ms
  express:router:route new /^(.+)$/ +0ms
  express:router:layer new /^(.+)$/ +0ms
  express:router:route post /^(.+)$/ +0ms
  express:router:layer new / +0ms
  express:router use / <anonymous> +0ms
  express:router:layer new / +0ms
  express:router dispatching GET /2939 +2s
  express:router query  : /2939 +1ms
  express:router expressInit  : /2939 +0ms
  express:router <anonymous>  : /2939 +0ms
  express:router serveStatic  : /2939 +0ms
  send stat "/webui/node_modules/ember-cli/node_modules/testem/public/2939" +2ms
  express:router jsonParser  : /2939 +1ms
  body-parser:json skip empty body +0ms
  express:router <anonymous>  : /2939 +0ms
  express:router logErrors  : /2939 +1ms
  express:router dispatching GET /2939/tests/index.html?coverage +6ms
  express:router query  : /2939/tests/index.html?coverage +0ms
  express:router expressInit  : /2939/tests/index.html?coverage +1ms
  express:router <anonymous>  : /2939/tests/index.html?coverage +1ms
  express:router serveStatic  : /2939/tests/index.html?coverage +0ms
  send stat "/webui/node_modules/ember-cli/node_modules/testem/public/2939/tests/index.html" +0ms
  express:router jsonParser  : /2939/tests/index.html?coverage +0ms
  body-parser:json skip empty body +9ms
  express:router <anonymous>  : /2939/tests/index.html?coverage +0ms
  express:router logErrors  : /2939/tests/index.html?coverage +0ms
  send stat "/webui/tmp/class-tests_dist-u4EqhXLW.tmp/tests/index.html" +2ms
  express:router <anonymous>  : /2939/tests/index.html?coverage +0ms
```

**Comment by jschilli** [comment]
@chrisvdp I'm stumped.  if you run ember test --server do the tests run to completion?

with ?coverage  without ?coverage query param

**Comment by chrisvdp** [comment]
`ember test --server` and `ember test` run to completion
`ember test -c ./testem.ci.json` without the `?coverage` param also runs to completion.

**Comment by chrisvdp** [comment]
Based on what I have seen the tests always run to completion, the problem is that something isn't exiting in ci mode with coverage. I get to this conclusion because I see valid output when I ctrl+c the process after letting it run for a while.

**Comment by jschilli** [comment]
so if you were to run in the browser with ?coverage and watch the network tab in chrome - does the call to write-coverage-data come back with a 200?

are other middlewares in use via addons?  

when DEBUG=* there is a list of addons registered.

given that we know ember-cli-blanket works in ci mode with mocha then the issue must be a specific interaction with another plugin.

if you can provide some isolated example it should be easy to find and fix.

Failing that, looking at addons or other middleware may reveal something.

**Comment by chrisvdp** [comment]
I ran phantom with `"phantomjs_debug_port": 5000` and found that it throws a 500 error. `Cannot read property 'map' of undefined`

![screen shot 2015-05-22 at 8 17 13 am](https://cloud.githubusercontent.com/assets/1924349/7772401/2c573384-005b-11e5-89b4-94b66fce4fcd.png)
 

---
What I wrote to answer your questions

1) A 200 does fire in chrome running with `ember test --server`
![screen shot 2015-05-22 at 8 10 35 am](https://cloud.githubusercontent.com/assets/1924349/7772285/7eb0477a-005a-11e5-8a2d-1b77f797edea.png)


2) This seems to be the addon output. I don't think that any addons are running in ci mode that aren't also running in `ember test --server`
```
  ember-cli:addons-factory  addons ordered as: +0ms [ 'tests-server-middleware',
  'ember-cli-content-security-policy',
  'history-support-middleware',
  'live-reload-middleware',
  'serve-files-middleware',
  'Ember CLI Blanket',
  'proxy-server-middleware',
  'Ember CLI Autoprefixer',
  'ember-cli-uglify',
  'broccoli-asset-rev',
  'ember-cli-app-version',
  'ember-cli-babel',
  'ember-cli-clean',
  'Ember CLI Coffeescript Addon',
  'ember-cli-dependency-checker',
  'ember-cli-htmlbars',
  'Ember CLI ic-ajax',
  'ember-cli-inline-content',
  'Ember CLI LESS',
  'Ember CLI Mocha',
  'Ember CLI moment',
  'ember-cli-pretender',
  'Ember CLI Simple Auth',
  'Ember CLI Simple Auth OAuth 2.0',
  'Ember CLI Simple Auth Testing',
  'ember-data',
  'ember-data-model-fragments',
  'ember-export-application-global',
  'Ember Localstorage Adapter',
  'ember-sinon',
  'ember-try' ]
  ember-cli:project addon: tests-server-middleware +1ms
  ember-cli:project addon: ember-cli-content-security-policy +1ms
  ember-cli:project addon: history-support-middleware +0ms
  ember-cli:project addon: live-reload-middleware +0ms
  ember-cli:project addon: serve-files-middleware +0ms
  ember-cli:project addon: Ember CLI Blanket +0ms
  ember-cli:project addon: proxy-server-middleware +0ms
  ember-cli:project addon: Ember CLI Autoprefixer +0ms
  ember-cli:project addon: ember-cli-uglify +0ms
  ember-cli:project addon: broccoli-asset-rev +0ms
  ember-cli:project addon: ember-cli-app-version +0ms
  ember-cli:project addon: ember-cli-babel +0ms
  ember-cli:project addon: ember-cli-clean +0ms
  ember-cli:project addon: Ember CLI Coffeescript Addon +0ms
  ember-cli:project addon: ember-cli-dependency-checker +0ms
  ember-cli:project addon: ember-cli-htmlbars +0ms
  ember-cli:project addon: Ember CLI ic-ajax +0ms
  ember-cli:project addon: ember-cli-inline-content +0ms
  ember-cli:project addon: Ember CLI LESS +0ms
  ember-cli:project addon: Ember CLI Mocha +0ms
  ember-cli:project addon: Ember CLI moment +0ms
  ember-cli:project addon: ember-cli-pretender +0ms
  ember-cli:project addon: Ember CLI Simple Auth +0ms
  ember-cli:project addon: Ember CLI Simple Auth OAuth 2.0 +0ms
  ember-cli:project addon: Ember CLI Simple Auth Testing +0ms
  ember-cli:project addon: ember-data +0ms
  ember-cli:project addon: ember-data-model-fragments +0ms
  ember-cli:project addon: ember-export-application-global +0ms
  ember-cli:project addon: Ember Localstorage Adapter +0ms
  ember-cli:project addon: ember-sinon +0ms
  ember-cli:project addon: ember-try +0ms
```

**Comment by chrisvdp** [comment]
stack trace of the map error.

```
TypeError: Cannot read property 'map' of undefined
  at Reporter.module.exports.Reporter.extend.transform (/webui/node_modules/ember-cli-blanket/lib/reporters/lcov-reporter.js:27:37)
  at Reporter.report (/webui/node_modules/ember-cli-blanket/lib/reporter.js:51:21)
  at /webui/node_modules/ember-cli-blanket/lib/coverage-middleware.js:39:16
  at Layer.handle [as handle_request] (/webui/node_modules/ember-cli/node_modules/express/lib/router/layer.js:82:5)
  at trim_prefix (/webui/node_modules/ember-cli/node_modules/express/lib/router/index.js:302:13)
  at /webui/node_modules/ember-cli/node_modules/express/lib/router/index.js:270:7
  at Function.proto.process_params (/webui/node_modules/ember-cli/node_modules/express/lib/router/index.js:321:12)
  at next (/webui/node_modules/ember-cli/node_modules/express/lib/router/index.js:261:10)
  at /webui/node_modules/ember-cli-blanket/node_modules/body-parser/lib/read.js:111:5
  at done (/webui/node_modules/ember-cli-blanket/node_modules/body-parser/node_modules/raw-body/index.js:248:14)
  at IncomingMessage.onEnd (/webui/node_modules/ember-cli-blanket/node_modules/body-parser/node_modules/raw-body/index.js:294:7)
  at IncomingMessage.g (events.js:199:16)
  at IncomingMessage.emit (events.js:104:17)
  at _stream_readable.js:908:16
  at process._tickCallback (node.js:355:11)
```

**Comment by jschilli** [comment]
I see that stack trace and can certainly put a check in for null coverage data coming back.

With that said, the refererer in the chrome case shows ?coverage whereas the example that returns 500 coverage was not turned on.

not sure what the difference is in invocation.

**Comment by chrisvdp** [comment]
Yes, you are right. Just a byproduct of playing with the configs. With coverage in phantomjs I do get a 200.
![screen shot 2015-05-22 at 8 33 22 am](https://cloud.githubusercontent.com/assets/1924349/7772774/80ceedec-005d-11e5-8ebc-f495241c32a7.png)

**Comment by chrisvdp** [comment]
Another interesting bit from that output is `{"suites":305,"tests":481,"passes":0,"pending":-9,"failures":490,"start":"2015-05-22T14:30:10.096Z","end":"2015-05-22T14:32:57.576Z"}`

**Comment by jschilli** [comment]
yeah, something’s borked there.

without a repo to repro in I’m just guessing.

**Comment by givanse** [comment]
I just ran into the same `TypeError: Cannot read property 'map' of undefined`.

I have a repo that you can play with.

    git clone -b cant-read-map git@github.com:givanse/mvc-tree.git

**Comment by jschilli** [comment]
I'll have a look. May be early next week. Thx

-jeff

**Comment by givanse** [comment]
So, I just "solved it". I was messing with the app name and ended up with a few mismatches. After correcting that, coverage worked fine again.

files to check:
* bower.json
* package.json
* config/environment.js
* tests/blanket-options.js
* app/index.html
* tests/index.html

**Comment by chrisvdp** [comment]
**Update** I spent most of yesterday debugging this. Findings were that tests hung upon running any integration/acceptance test. I tried to init a (new ember-cli project)[https://github.com/chrisvdp/ember-test-blanket-integration] to recreate it, but was unable to. If running with `ember test --server` the tests should run and the results should be displayed with the coverage results below, In the case of acceptance tests, the results aren't displayed and the coverage results override everything. I assume this is the source of the issue, but can't find a definitive cause. 

As far as I can tell this issue stems from importing the [quint version of blanket](https://github.com/sglanzer/ember-cli-blanket/blob/master/index.js#L29) no matter the framework specified in the testem config file.

**Comment by chrisvdp** [comment]
PS. Thanks for all the help @jschilli !! :balloon:

**Comment by NullVoxPopuli** [comment]
you can reproduce in this repo: https://github.com/NullVoxPopuli/aeonvera-ui 

:-(

**Comment by devinus** [comment]
I'm getting the exact same error:

```
TypeError: Cannot read property 'map' of undefined
    at Reporter.module.exports.Reporter.extend.transform (/Users/devinus/Projects/rebelmail-dashboard/node_modules/ember-cli-blanket/lib/reporters/lcov-reporter.js:40:37)
    at Reporter.report (/Users/devinus/Projects/rebelmail-dashboard/node_modules/ember-cli-blanket/lib/reporter.js:51:21)
    at /Users/devinus/Projects/rebelmail-dashboard/node_modules/ember-cli-blanket/lib/coverage-middleware.js:37:16
    at arrayEach (/Users/devinus/Projects/rebelmail-dashboard/node_modules/ember-cli-blanket/node_modules/lodash/index.js:1289:13)
    at Function.<anonymous> (/Users/devinus/Projects/rebelmail-dashboard/node_modules/ember-cli-blanket/node_modules/lodash/index.js:3345:13)
    at /Users/devinus/Projects/rebelmail-dashboard/node_modules/ember-cli-blanket/lib/coverage-middleware.js:36:7
    at Layer.handle [as handle_request] (/Users/devinus/Projects/rebelmail-dashboard/node_modules/express/lib/router/layer.js:95:5)
    at next (/Users/devinus/Projects/rebelmail-dashboard/node_modules/express/lib/router/route.js:131:13)
    at /Users/devinus/Projects/rebelmail-dashboard/node_modules/body-parser/lib/read.js:129:5
    at invokeCallback (/Users/devinus/Projects/rebelmail-dashboard/node_modules/raw-body/index.js:262:16)
    at done (/Users/devinus/Projects/rebelmail-dashboard/node_modules/raw-body/index.js:251:7)
    at IncomingMessage.onEnd (/Users/devinus/Projects/rebelmail-dashboard/node_modules/raw-body/index.js:308:7)
    at emitNone (events.js:80:13)
    at IncomingMessage.emit (events.js:179:7)
    at endReadableNT (_stream_readable.js:906:12)
    at nextTickCallbackWith2Args (node.js:474:9)
    at process._tickDomainCallback (node.js:429:17)
```

**Comment by jschilli** [comment]
I believe this was resolved with #111 

closing for now - feel free to reopen and/or comment if still an issue
